iss#6: implemented logic to determine how skywire-updater finds config file.

atang152 opened the pull request:
Implement services to find config file path based on the following order:
1) CLI argument, 
2) env, 
3) List of default path: i) working directory, ii) home (.skycoin/skywire/skywire-config.json), iii) local (/usr/local/skycoin/skywire/skywire-config.json). 

Code is based on library from Skywire mainnet's internal/pathutil/configpath.go. 'Invalid command' was returned when using the original skywire-update [config-path] code. It turns out that adding subcommands prevents RootCmd from accepting path arguments. Therefore, 'updateCmd' was created so update services can take in an optional path argument. 

First contribution. Not sure if it's done correctly but I hope it helps. Thank you.

atang152 committed:
issue#6 - Implement services to find a config file path based on the following order 1) CLI argument, 2) ENV, 3) List of default path - i) working directory, home (.skycoin/skywire/skywire-config.json), iii) local (/usr/local/skycoin/skywire/skywire-config.json). Code base is based on library from  Skywire mainnet's internal/pathutil/configpath.go.

atang152 committed:
Added preliminary unit test- configpath_test.go

atang152 committed:
minor changes to configpath_test.go. Only testing one case where default config path is used. Coverage for test is low.

atang152 committed:
issue#6 - Refractor skywire-update. 'Invalid command' was returned when using skywire-update [config-path]. It turns out that adding cobra subcommands prevents RootCmd from accepting path arguments. Therefore, 'updateCmd' was created so update services can take an optional argument. Update will be based on config path loaded from the following order: 1) CLI argument 2) env 3) list of default path.

atang152 committed:
Merge branch 'develop_v2' into develop

"merged develop_v2 to develop_v1. issue#6 - Refractor skywire-update. 'Invalid command' was returned when using skywire-update [config-path]. It turns out that adding cobra subcommands prevents RootCmd from accepting path arguments. Therefore, 'updateCmd' was created so update services can take an optional argument. Update will be based on config path loaded from the following order: 1) CLI argument 2) env 3) list of default path. "

evanlinjin commented on README.md:
```diff
@@ -8,28 +8,20 @@
-
-  1. /Users/anonymous/.skycoin/skywire-updater/config.yml
-  2. /usr/local/skycoin/skywire-updater/config.yml
+associated with skywire. Services to be updated will be based on a specified configuration file.
```
Please reword to: `... based on specified configuration files.`

atang152 replied:
Apologies for the late reply. Resolved.

evanlinjin commented on README.md:
```diff
@@ -8,28 +8,20 @@
+associated with skywire. Services to be updated will be based on a specified configuration file.
 
 Usage:
-  skywire-updater [config-path] [flags]
```
I would prefer it if we read the `[config-file]` via a CLI flag, instead of a new sub-command.  😜

atang152 replied:
Thank you! I was thinking to use CLI flags originally too; I will go back and review the code again and see if I could make this change. Thank you!

atang152 replied:
Hi Evan. I made a change to the code. It can now reads in [config-file] via CLI flag instead of having to type a new Cobra subcommand. It turns out that by default Cobra only parses local flags on the target command, any local flags on parent commands are ignored. So by enabling this argument Command.TraverseChildren allows Cobra to parse local flags on each command before executing the target command. Please let me know if this logic is okay. Thanks!

evanlinjin commented on cmd/skywire-updater/commands/gen-config.go:
```diff
@@ -6,11 +6,10 @@ import (
-	"github.com/spf13/cobra"
-	"gopkg.in/yaml.v2"
-
 	"github.com/skycoin/skywire-updater/internal/pathutil"
```
We should probably make this module public in `skycoin/skywire`. (Our bad!)

atang152 replied:
Moved pathutil from /internal/pathutil to /pkg/util/pathutil as per https://github.com/skycoin/skywire/pull/295

evanlinjin commented on cmd/skywire-updater/commands/gen-config.go:
```diff
@@ -6,11 +6,10 @@ import (
 	"github.com/skycoin/skywire-updater/internal/pathutil"
 	"github.com/skycoin/skywire-updater/pkg/update"
+	"github.com/spf13/cobra"
+	"gopkg.in/yaml.v2"
```
This does not look `goimported`?

atang152 replied:
Hi Evan. Can you clarify a bit more here? Do you mean that these imports shouldn't be there? I think it could be because I am having issue with my goimported. golangcli-lint installation fails for me whenever I run make install-linters. I am not able to narrow down the reason at the moment but will investigate further. Thanks!

evanlinjin replied:
@atang152 have you tried installing with env `GO111MODULE=off`?

evanlinjin replied:
@atang152 to format the code:

```bash
$ goimports -w -local github.com/skycoin/skywire-updater ./cmd
$ goimports -w -local github.com/skycoin/skywire-updater ./pkg
```

(Using https://github.com/skycoin/skywire/blob/c3a15be793212133c74259cda857fbfa9efb7df2/Makefile#L85-L88 as reference.)

evanlinjin replied:
The `Makefile` is outdated (as you know) and doesn't format the `/cmd` folder.

evanlinjin commented on cmd/skywire-updater/commands/update.go:
```diff
@@ -0,0 +1,69 @@
+
+const configEnv = "SW_UPDATER_CONFIG"
+
+var defaultConfigPaths = [3]string{
```
I would prefer:
```golang
var defaultPaths = pathutil.UpdaterDefaults()

/* ... */
// 3. %s`, defaultPaths[pathutil.WorkingDirLoc], defaultPaths[pathutil.HomeLoc], defaultPaths[pathutil.LocalLoc]),

```

atang152 replied:
Thank you for bringing this up. Resolved.

evanlinjin reviewed (changes requested):
Great work on a first PR! Code is very clean and excellent work in general.

However, it would be appreciated if you implement the changes as below.

Additionally, if you can do a PR in `skycoin/skywire` to move `/internal/pathutil` to `/pkg/util/pathutil`, and update this PR accordingly, that will be a plus.

Please contact me on Telegram `@evanlinjin` regardless.

ivcosla commented on internal/pathutil/configpath.go:
```diff
@@ -0,0 +1,159 @@
+type ConfigPaths map[ConfigLocationType]string
+
+// String implements fmt.Stringer for ConfigPaths.
+func (dp ConfigPaths) String() string {
```
Just one question, why did you chose to call this `dp`?

atang152 replied:
Hi Ivan. I believe dp is "defaultPath". The original configpath.go code was actually written by Evan to manage the configpath for mainnet (https://github.com/skycoin/skywire/blob/mainnet/internal/pathutil/configpath.go).  I re-used Evan's code and incorporated a function that is specific for Skywire-Updater config. I hope that this doesn't cause any issues. Thank you.

evanlinjin replied:
@atang152 What is the additional function you have implemented? Can we make use of the version in `skycoin/skywire` now?

atang152 replied:
Hi Evan. In your configPath.go code, you created two funcs that returns default path for 1) skywire-node and 2) manager node. They are called "NodeDefaults()" and "ManagerDefaults" respectively. I added one extra func called UpdaterDefaults() that simply returns the default config paths specifically for Skywire-Updater. 

_Can we make use of the version in skycoin/skywire now?_ : It can be incorporated into the current version in skycoin/skywire but I don't think it will do anything since the current Skywire-Updater is independent and seperated from the Skywire-Updater. Please let me know if there is anything I should do. Thanks!

evanlinjin replied:
@atang152 The changes that are required are as follows.
1. Import `github.com/skycoin/skywire/pkg/util/pathutil`, rather than copying the source to `github.com/skycoin/skywire-updater/...`.

    The command to achieve this (with env `GO111MODULE=on`) is: 
    ```golang
    go get github.com/skycoin/skywire/pkg/util/pathutil@mainnet
    ```
    Then you can import this in the code.

2. Please put `UpdaterDefaults()` in a more suitable location (in `/cmd/skywire-updater/root.go` is fine).

evanlinjin replied:
@atang152 https://github.com/golang/go/wiki/modules

Hopefully this resource will help.

evanlinjin replied:
> @atang152 The changes that are required are as follows.
> 
> 1. Import `github.com/skycoin/skywire/pkg/util/pathutil`, rather than copying the source to `github.com/skycoin/skywire-updater/...`.
>    The command to achieve this (with env `GO111MODULE=on`) is:
>    ```go
>    go get github.com/skycoin/skywire/pkg/util/pathutil@mainnet
>    ```
>    
>    
>    Then you can import this in the code.
> 2. Please put `UpdaterDefaults()` in a more suitable location (in `/cmd/skywire-updater/root.go` is fine).

@atang152 How is progress with this? Just referencing this here as I see the source comment is outdated.

https://github.com/golang/go/wiki/modules

ivcosla reviewed (approved):
This is definitely a very good job!, congrats.

atang152 committed:
[WIP] fixing pr#7. make install-linters seems to fail  with go1.11 and above.

evanlinjin commented:
@atang152 in the future, I would name the branch `feature/update-config-logic` or something similar.

atang152 committed:
issue#6 - fixed changes requested from code review for pr#7.
1) Refractor code to now read in [config-file] via CLI flag instead of using a new Cobra subcommand (As per https://github.com/skycoin/skywire-updater/pull/7#discussion_r275936202). Previous issue of not being able to accept flag arguments due to addition of subcommand was resolved by enabling Command.TraverseChildren. Enabling TraverseChildren allows Cobra to parse local flags on each command before executing target subcommands.
2) Moved /internal/pathutil to /pkg/util/pathutil

atang152 committed:
minor changes to .gitignore to exclude .vscode

atang152 committed:
minor changes to .gitignore to exclude .vscode

evanlinjin commented:
https://github.com/skycoin/skywire-updater/pull/7#discussion_r276609705
> @atang152 The changes that are required are as follows.
> 
> 1. Import `github.com/skycoin/skywire/pkg/util/pathutil`, rather than copying the source to `github.com/skycoin/skywire-updater/...`.
>    The command to achieve this (with env `GO111MODULE=on`) is:
>    ```go
>    go get github.com/skycoin/skywire/pkg/util/pathutil@mainnet
>    ```
>    
>    
>    Then you can import this in the code.
> 2. Please put `UpdaterDefaults()` in a more suitable location (in `/cmd/skywire-updater/root.go` is fine).

@atang152 How is progress with this? Just referencing this here as I see the source comment is outdated.

https://github.com/golang/go/wiki/modules

atang152 committed:
Fixed import and code to read in [config] via CLI

1) Fixed import path by importing github.com/skycoin/skywire/pkg/util/pathutil, rather than copying the source to github.com/skycoin/skywire-updater/... Done so by using GO111MODULE=on go get github.com/skycoin/skywire/pkg/util/pathutil@mainnet
2) Refractor code to now read in [config-file] via CLI flag instead of using a new Cobra subcommand (As per https://github.com/skycoin/skywire-updater/pull/7#discussion_r275936202). Previous issue of not being able to accept flag arguments due to addition of subcommand was resolved by enabling Command.TraverseChildren. Enabling TraverseChildren allows Cobra to parse local flags on each command before executing target subcommands.

atang152 committed:
Delete pathutil/homedir path & ran go mod vendor

1) Removed pkg/util/pathutil from src and import it directly from skycoin/skywire instead.
2) Ran GO111MODULE=on go mod vendor to update vendor directory to include all packages needed to build/test based on go.mod and go src code.

evanlinjin commented on cmd/skywire-updater/commands/root.go:
```diff
@@ -1,65 +1,57 @@
-			configPath = args[0]
-		}
+
+		configPath := pathutil.FindConfigPath(args, 0, configEnv, UpdaterDefaults())
```
There is no need to run `UpdaterDefaults()` again. We can just use the `defaultPaths` variable.

atang152 committed:
Ran goimports and minor edit to Makefile

Ran goimports with the correct path. Editted Makefile to match Skywire standardize format.

evanlinjin reviewed (approved):
Good work!
